src/_cffi_src/openssl/src/osrandom_engine.c

```diff
@@ -0,0 +1,429 @@
+                n = close(fd);
+            } while (n < 0 && errno == EINTR);
+            return urandom_cache.fd;
+        }
+        urandom_cache.st_dev = st.st_dev;
+        urandom_cache.st_ino = st.st_ino;
+        urandom_cache.fd = fd;
+    }
+    return urandom_cache.fd;
+
+  error:
+    if (fd != -1) {
+        do {
+            n = close(fd);
+        } while (n < 0 && errno == EINTR);
+    }
+    CRYPTOGRAPHY_OSRANDOM_put_error(
+        "osrandom_engine.py:dev_urandom_fd()");
+    return -1;
+}
+
+static int dev_urandom_read(unsigned char *buffer, int size) {
+    int fd;
+    ssize_t n;
+
+    fd = dev_urandom_fd();
+    if (fd < 0) {
+        return 0;
+    }
+
+    while (size > 0) {
+        do {
+            n = read(fd, buffer, (size_t)size);
+        } while (n < 0 && errno == EINTR);
+
+        if (n <= 0) {
+            CRYPTOGRAPHY_OSRANDOM_put_error(
+                "osrandom_engine.py:dev_urandom_read()");
+            return 0;
+        }
+        buffer += n;
+        size -= n;
+    }
+    return 1;
+}
+
+static void dev_urandom_close(void) {
+    if (urandom_cache.fd >= 0) {
+        int fd, n;
+        struct stat st;
+
+        if (fstat(urandom_cache.fd, &st)
+                && st.st_dev == urandom_cache.st_dev
+                && st.st_ino == urandom_cache.st_ino) {
+            fd = urandom_cache.fd;
+            urandom_cache.fd = -1;
+            do {
+                n = close(fd);
+            } while (n < 0 && errno == EINTR);
+        }
+    }
+}
+#endif /* CRYPTOGRAPHY_OSRANDOM_NEEDS_DEV_URANDOM */
+
+/****************************************************************************
+ * Linux getrandom engine with fallback to dev_urandom
+ */
+
+#if CRYPTOGRAPHY_OSRANDOM_ENGINE == CRYPTOGRAPHY_OSRANDOM_ENGINE_GETRANDOM
+static const char *Cryptography_osrandom_engine_name = "osrandom_engine getrandom()";
+
+static int getrandom_works = -1;
+
+static int osrandom_init(ENGINE *e) {
+    /* We try to detect working getrandom until we succeed. */
+    if (getrandom_works != 1) {
+        long n;
+        char dest[1];
+        n = syscall(SYS_getrandom, dest, sizeof(dest), GRND_NONBLOCK);
+        if (n < 0) {
+            /* Can fail with:
+             * - ENOSYS: Kernel does not support the syscall.
+             * - ENOPERM: seccomp prevents syscall.
+             * - EAGAIN: Kernel CRPNG has not been seeded yet.
+             * EINTR cannot occur for buflen < 256.
+             */
+            getrandom_works = 0;
+        } else {
+            getrandom_works = 1;
+        }
+    }
+    /* fallback to dev urandom */
+    if (getrandom_works == 0) {
+        int fd = dev_urandom_fd();
+        if (fd < 0) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static int osrandom_rand_bytes(unsigned char *buffer, int size) {
+    if (getrandom_works == 1) {
+        long n;
+        while (size > 0) {
+            do {
+                n = syscall(SYS_getrandom, buffer, size, GRND_NONBLOCK);
```

Comment:
http://man7.org/linux/man-pages/man2/getrandom.2.html

>    If the
>       /dev/urandom pool has been initialized, reads of up to 256 bytes will
>       always return as many bytes as requested and will not be interrupted
>       by signals.  No such guarantees apply for larger buffer sizes.  For
>       example, if the call is interrupted by a signal handler, it may
>       return a partially filled buffer, or fail with the error EINTR.